kafka-event-listener/src/test/java/org/hypertrace/core/kafka/event/listener/KafkaEventListenerTest.java

```diff
@@ -0,0 +1,105 @@
+package org.hypertrace.core.kafka.event.listener;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.mockito.Mockito.mock;
+
+import com.github.benmanes.caffeine.cache.AsyncLoadingCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.typesafe.config.ConfigFactory;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.MockConsumer;
+import org.apache.kafka.clients.consumer.OffsetResetStrategy;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.junit.jupiter.api.Test;
+
+class KafkaEventListenerTest {
+
+  @Test
+  void testEventModificationCache() throws Exception {
+    EventModificationCache eventModificationCache = new EventModificationCache();
+    // kafka consumer mock setup
+    MockConsumer<String, Long> kafkaConsumer = new MockConsumer<>(OffsetResetStrategy.LATEST);
+    String topic = "event-update-topic";
+    kafkaConsumer.updatePartitions(
+        topic,
+        List.of(
+            getPartitionInfo(topic, 0),
+            getPartitionInfo(topic, 1),
+            getPartitionInfo(topic, 2),
+            getPartitionInfo(topic, 3)));
+    HashMap<TopicPartition, Long> endOffsets = new HashMap<>();
+    endOffsets.put(new TopicPartition(topic, 0), 50L);
+    endOffsets.put(new TopicPartition(topic, 1), 50L);
+    endOffsets.put(new TopicPartition(topic, 2), 50L);
+    endOffsets.put(new TopicPartition(topic, 3), 50L);
+    kafkaConsumer.updateEndOffsets(endOffsets);
+    // register consumer
+    eventModificationCache.registerConsumer(
+        new KafkaEventListenerConsumer(
+            "modification-event-consumer",
+            ConfigFactory.parseMap(Map.of("topic.name", topic, "poll.timeout", "9ms")),
+            kafkaConsumer,
+            eventModificationCache));
+    Thread.sleep(10);
+    assertEquals(10L, eventModificationCache.get(10));
+    assertEquals(100L, eventModificationCache.get(100));
+    // not present key won't trigger any population
+    kafkaConsumer.addRecord(new ConsumerRecord<>(topic, 0, 100, "32", 89L));
+    Thread.sleep(10);
+    assertFalse(eventModificationCache.hasKey(32));
+    // existing key will be modified based on entry
+    kafkaConsumer.addRecord(new ConsumerRecord<>(topic, 3, 200, "10", -3L));
+    Thread.sleep(10);
```

Comment:
The clock is available for kafka streams based applications where you have wall clock advance. Here it is a plain consumer and in test we use MockConsumer.

Here the sleep is required because of poll timeout mechanism and mock consumer reacting to add record. If our while loop is already in prev poll then the new record added won't show up until prev poll times out. Hence we need a Thread.sleep, and timeout was 9ms, and sleep has been put as 10ms

```java
do {
      try {
        ConsumerRecords<D, C> records = kafkaConsumer.poll(pollTimeout);
```

Reply:
Not super familiar with the test constructs they give us, it surprises me they don't give the test consumer a way to advance the clock and avoid the sleep.